src/xform.c

```diff
@@ -1142,12 +1151,56 @@ resize_stream(Gif_Stream* gfs,
         && method != SCALE_METHOD_LANCZOS3 && method != SCALE_METHOD_MITCHELL)
         method = SCALE_METHOD_POINT;
 
-    for (sctx.imageno = 0; sctx.imageno < gfs->nimages; ++sctx.imageno) {
-        sctx.gfi = gfs->images[sctx.imageno];
-        scale_image(&sctx, method);
+#ifdef ENABLE_THREADED
+    pthread_mutex_init(&kd3_sort_lock, 0);
+#endif
+
+    int i;
+    if (resize_threads > 0) {
+#ifdef ENABLE_THREADED
+        scale_thread_context* contexts = malloc(gfs->nimages * sizeof(scale_thread_context));
+        pthread_t * pthreads = malloc(gfs->nimages * sizeof(pthread_t));
+
+        for (i = 0; i < gfs->nimages;) {
+            int start = i;
+            int end = min(i + resize_threads, gfs->nimages);
+            for (; i < end; ++i) {
+                sctx_init(&(contexts[i].sctx), gfs, nw, nh);
+                contexts[i].sctx.imageno = i;
+                contexts[i].sctx.scale_colors = scale_colors;
+                contexts[i].sctx.gfi = gfs->images[i];
+                contexts[i].method = method;
+
+                /* Create worker thread */
+                pthread_create(&pthreads[i], NULL, scale_image_threaded, &contexts[i]);
+            }
+            int j = start;
+            for (; j < end; ++j) {
+                pthread_join(pthreads[j], NULL);
+            }
+        }
```

Comment:
It would be just a tiny bit more work to add a true “work stealing” implementation and I think that would be a nice benefit!

Something like this:

```
typedef struct {
    pthread_t threadid;
    Gif_Stream* gfs;
    int imageno;
    int* next_imageno;
    int nw;
    int nh;
    ... scale_colors;
    ... method;
} scale_thread_context;

void* scale_image_threaded(void* args) {
    scale_thread_context* ctx = (scale_thread_context*) args;
    scale_context sctx;
    sctx_init(&sctx, ctx->gfs, ctx->nw, ctx->nh);
    sctx.scale_colors = ctx->scale_colors;
    do {
        sctx.imageno = ctx->imageno;
        sctx.gfi = ctx->gfs->images[ctx->imageno];
        scale_image(&sctx, ctx->method);
        ctx->imageno = __sync_add_and_fetch(ctx->next_imageno, 1);
    } while (ctx->imageno < ctx->gfs->nimages);
    sctx_cleanup(&sctx);
    return 0;
}

... in scale_image ...

    int nthreads = min(resize_threads, gfs->nimages);
    scale_thread_context* ctx = Gif_NewArray(...);
    int next_imageno = nthreads;
    for (int i = 0; i < nthreads; ++i) {
          /* set up ctx[i] */ ...
          ctx[i].imageno = i;
          ctx[i].next_imageno = &next_imageno;
          pthread_create(&ctx[i].threadid, ...);
    }
    for (int i = 0; i < nthreads; ++i) pthread_join ... ;
```

Make sense?